File: junit/rails/game/state/StringStateTest.java
package rails.game.state;


import static org.fest.assertions.api.Assertions.assertThat;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

import org.junit.Before;
import org.junit.Test;

public class StringStateTest {

    private final static String DEFAULT_ID = "Default";
    private final static String INIT_ID = "Init"; 
    private final static String INIT = "INIT";
    private final static String OTHER = "OTHER";

    private Root root;
    
    private StringState stateDefault;
    private StringState stateInit;
    

    @Before
    public void setUp() {
        root = StateTestUtils.setUpRoot();
        
        
        stateDefault = StringState.create(root, DEFAULT_ID);
        stateInit = StringState.create(root, INIT_ID, INIT);
    }
    
    @Test
    public void testValue() {
        assertEquals(stateDefault.value(), null);
        assertEquals(stateInit.value(), INIT);
    }

    @Test
    public void testSet() {
        stateDefault.set(OTHER);
        assertEquals(stateDefault.value(), OTHER);
        stateInit.set("");
        assertEquals(stateInit.value(), "");
        stateInit.set(null);
        assertEquals(stateInit.value(), null);
    }

    @Test
    public void testAppend() {
        stateDefault.append(OTHER, null);
        assertEquals(stateDefault.value(), OTHER);
        stateDefault.append(OTHER, "");
        assertEquals(stateDefault.value(), OTHER + OTHER);
        
        stateInit.append(OTHER, ",");
        assertEquals(stateInit.value(), INIT + "," + OTHER);
    }
    
    
    @Test
    public void testSetSameIgnored() {
        stateDefault.set(null);
        stateInit.set(null);
        StateTestUtils.close(root);
        assertThat(StateTestUtils.getLastClosedChangeSet(root).getObservableStates()).doesNotContain(stateDefault);
        assertThat(StateTestUtils.getLastClosedChangeSet(root).getObservableStates()).contains(stateInit);

        StateTestUtils.newChangeSet(root);
        stateDefault.set("");
        stateInit.set(null);
        StateTestUtils.close(root);
        assertThat(StateTestUtils.getLastClosedChangeSet(root).getObservableStates()).contains(stateDefault);
        assertThat(StateTestUtils.getLastClosedChangeSet(root).getObservableStates()).doesNotContain(stateInit);
    }

    @Test
    public void testUndoRedo() {
        
        stateDefault.set(OTHER);
        stateDefault.append(OTHER, null);
        
        stateInit.append(OTHER, "");
        stateInit.set(null);
        StateTestUtils.close(root);
        
        assertEquals(stateDefault.value(), OTHER+OTHER);
        assertNull(stateInit.value());

        StateTestUtils.undo(root);
        assertEquals(stateDefault.value(), null);
        assertEquals(stateInit.value(), INIT);

        StateTestUtils.redo(root);
        assertEquals(stateDefault.value(), OTHER+OTHER);
        assertNull(stateInit.value());
    }


}

File: dotnet/GameStateTest/StringStateTest.cs
using GameLib.Net.Game.State;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GameStateTest
{
    [TestClass]
    public class StringStateTest
    {
        private const string DEFAULT_ID = "Default";
        private const string INIT_ID = "Init";
        private const string INIT = "INIT";
        private const string OTHER = "OTHER";

        private Root root;

        private StringState stateDefault;
        private StringState stateInit;


        [TestInitialize]
        public void SetUp()
        {
            root = StateTestUtils.SetUpRoot();

            stateDefault = StringState.Create(root, DEFAULT_ID);
            stateInit = StringState.Create(root, INIT_ID, INIT);
        }

        [TestMethod]
        public void TestValue()
        {
            Assert.AreEqual(stateDefault.Value, null);
            Assert.AreEqual(stateInit.Value, INIT);
        }

        [TestMethod]
        public void TestSet()
        {
            stateDefault.Set(OTHER);
            Assert.AreEqual(stateDefault.Value, OTHER);
            stateInit.Set("");
            Assert.AreEqual(stateInit.Value, "");
            stateInit.Set(null);
            Assert.AreEqual(stateInit.Value, null);
        }

        [TestMethod]
        public void TestAppend()
        {
            stateDefault.Append(OTHER, null);
            Assert.AreEqual(stateDefault.Value, OTHER);
            stateDefault.Append(OTHER, "");
            Assert.AreEqual(stateDefault.Value, OTHER + OTHER);

            stateInit.Append(OTHER, ",");
            Assert.AreEqual(stateInit.Value, INIT + "," + OTHER);
        }


        [TestMethod]
        public void TestSetSameIgnored()
        {
            stateDefault.Set(null);
            stateInit.Set(null);
            StateTestUtils.Close(root);
            //assertThat(StateTestUtils.getPreviousChangeSet(root).getStates()).doesNotContain(stateDefault);
            Assert.IsTrue(!StateTestUtils.GetPreviousChangeSet(root).GetStates().Contains(stateDefault));
            //assertThat(StateTestUtils.getPreviousChangeSet(root).getStates()).contains(stateInit);
            Assert.IsTrue(StateTestUtils.GetPreviousChangeSet(root).GetStates().Contains(stateInit));

            StateTestUtils.Close(root);
            stateDefault.Set("");
            stateInit.Set(null);
            StateTestUtils.Close(root);
            //assertThat(StateTestUtils.getPreviousChangeSet(root).getStates()).contains(stateDefault);
            Assert.IsTrue(StateTestUtils.GetPreviousChangeSet(root).GetStates().Contains(stateDefault));
            //assertThat(StateTestUtils.getPreviousChangeSet(root).getStates()).doesNotContain(stateInit);
            Assert.IsTrue(!StateTestUtils.GetPreviousChangeSet(root).GetStates().Contains(stateInit));
        }

        [TestMethod]
        public void TestUndoRedo()
        {

            stateDefault.Set(OTHER);
            stateDefault.Append(OTHER, null);

            stateInit.Append(OTHER, "");
            stateInit.Set(null);
            StateTestUtils.Close(root);

            Assert.AreEqual(stateDefault.Value, OTHER + OTHER);
            Assert.IsNull(stateInit.Value);

            StateTestUtils.Undo(root);
            Assert.AreEqual(stateDefault.Value, null);
            Assert.AreEqual(stateInit.Value, INIT);

            StateTestUtils.Redo(root);
            Assert.AreEqual(stateDefault.Value, OTHER + OTHER);
            Assert.IsNull(stateInit.Value);
        }

    }
}
